### PR TITLE
rodeos: retry on bind failure

### PR DESCRIPTION
**Pull request by tbfleming**
## Change Type
**Select ONE**
- [ ] Documentation

- [ ] Stability bug fix

- [x] Other

- [ ] Other - special case

Sometimes when rodeos starts up, the OS fails the socket bind saying the port is already in use. The OS then allows it after some unknown amount of time. We've seen this behavior on OSX after killing (gracefully) the process then restarting it. We've also seen this behavior when starting containers on k8s.

rodeos now automatically retries the bind, waiting ~1s between each try. A new option, `--wql-retries` limits the number of retries (0 = no retrying, default = no limit). rodeos shuts down if it hits the retry limit.

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

**Comment by spoonincode** [comment]
Did you try `SO_REUSEADDR`?

```c++
boost::asio::socket_base::reuse_address option(true);
acceptor.set_option(option);
```

**Comment by tbfleming** [comment]
Ick. I confused that option with `SO_REUSEPORT`; I didn't want to open a vulnerability with that.

**Comment by tbfleming** [comment]
Those options are poorly named.

File: programs/rodeos/wasm_ql_plugin.cpp [hunk 1]
```diff
@@ -17,14 +17,45 @@ namespace b1 {
...
+         timer.async_wait([this](auto&) {
+            ilog("retry...");
+            try {
+               try {
+                  start_http();
+               }
+               FC_LOG_AND_RETHROW()
+            } catch (...) {
+               elog("shutting down");
+               app().shutdown();
```
Comment (swatanabe-b1) [inline programs/rodeos/wasm_ql_plugin.cpp:47]: Directly calling `app().shutdown()` is dangerous.  It's better to call `app().quit()` and let appbase break out of the event loop normally.

Comment (heifner) [reply programs/rodeos/wasm_ql_plugin.cpp:47]: We really should make that method private.

Comment (tbfleming) [reply programs/rodeos/wasm_ql_plugin.cpp]: Yep. And switch http_plugin to using quit.

Comment (heifner) [reply programs/rodeos/wasm_ql_plugin.cpp]: I saw that. I'll create a PR to develop for it.

**Comment by spoonincode** [comment]
Do you really need to keep the loop around still?

**Comment by tbfleming** [comment]
Probably not. The round-trip-time for customer testing is so high that I prefer to leave it in just in case...

**Pull request merged**